src/Storages/StorageHudi.cpp

```diff
@@ -0,0 +1,242 @@
+        columns_,
+        constraints_,
+        comment,
+        context_,
+        std::nullopt);
+}
+
+Pipe StorageHudi::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum processed_stage,
+    size_t max_block_size,
+    unsigned num_streams)
+{
+    StorageS3::updateS3Configuration(context, base_configuration);
+    return s3engine->read(column_names, storage_snapshot, query_info, context, processed_stage, max_block_size, num_streams);
+}
+
+std::vector<std::string> StorageHudi::getKeysFromS3()
+{
+    std::vector<std::string> keys;
+
+    const auto & client = base_configuration.client;
+
+    Aws::S3::Model::ListObjectsV2Request request;
+    Aws::S3::Model::ListObjectsV2Outcome outcome;
+
+    bool is_finished{false};
+    const auto bucket{base_configuration.uri.bucket};
+
+    request.SetBucket(bucket);
+    request.SetPrefix(table_path);
+
+    while (!is_finished)
+    {
+        outcome = client->ListObjectsV2(request);
+        if (!outcome.IsSuccess())
+            throw Exception(
+                ErrorCodes::S3_ERROR,
+                "Could not list objects in bucket {} with key {}, S3 exception: {}, message: {}",
+                quoteString(bucket),
+                quoteString(table_path),
+                backQuote(outcome.GetError().GetExceptionName()),
+                quoteString(outcome.GetError().GetMessage()));
+
+        const auto & result_batch = outcome.GetResult().GetContents();
+        for (const auto & obj : result_batch)
+        {
+            const auto & filename = obj.GetKey().substr(table_path.size()); // object name without tablepath prefix
+            keys.push_back(filename);
+            LOG_DEBUG(log, "Found file: {}", filename);
+        }
+
+        request.SetContinuationToken(outcome.GetResult().GetNextContinuationToken());
+        is_finished = !outcome.GetResult().GetIsTruncated();
+    }
+
+    return keys;
+}
+
+std::string StorageHudi::generateQueryFromKeys(std::vector<std::string> && keys, String format)
+{
+    // make format lowercase
+    std::transform(format.begin(), format.end(), format.begin(), [](unsigned char c) { return std::tolower(c); });
+
+    // filter only files with specific format
+    std::erase_if(keys, [&format](const std::string & s) { return std::filesystem::path(s).extension() != "." + format; });
+
+    // for each partition path take only latest file
+
+    std::unordered_map<std::string, std::pair<std::string, uint64_t>> latest_parquets;
+
+    for (const auto & key : keys)
+    {
+        auto slash = key.find_last_of("/");
+        std::string path;
+        if (slash == std::string::npos)
+        {
+            path = "";
+        }
+        else
+        {
+            path = key.substr(0, slash);
+        }
+
+        // every filename contains metadata split by "_", timestamp is after last "_"
```

Comment:
please add a comment why do we add latest by timestamp files? and what kind of files it is? metadata files?

Reply:
Can you add this comment inside the code, please? It should be understandable from code what is going on, when a person does not know internal of hudi/delta lake. Add same for other places, where I asked to explain "why", please :)

Reply:
Sure, will add comments.